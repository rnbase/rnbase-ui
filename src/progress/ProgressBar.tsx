import React, { useState, useEffect } from 'react';
import { Animated, Easing, StyleProp, StyleSheet, View, ViewProps, ViewStyle } from 'react-native';

import { Themed, Theme, WithThemeProps, withTheme } from '../theming';
import { getRadius, useLayout } from '../helpers';

interface ProgressBarProps extends ViewProps {
  size?: number;
  cornerRadius?: number | string;
  value?: number;
  style?: StyleProp<ViewStyle>;
  indicatorStyle?: StyleProp<ViewStyle>;
}

const ProgressBar: React.FC<Themed<typeof createStyleSheet, ProgressBarProps>> = ({
  Colors: _Colors,
  styles,
  value,
  size = 6,
  cornerRadius = '50%',
  style,
  indicatorStyle,
  ...props
}) => {
  const [{ width }, onLayout] = useLayout();
  const [animatedValue] = useState(() => new Animated.Value(value || 0));

  useEffect(() => {
    if (value !== undefined) {
      Animated.timing(animatedValue, {
        duration: 200,
        toValue: value,
        useNativeDriver: true,
        easing: Easing.out(Easing.cubic),
      }).start();
    } else {
      animatedValue.setValue(0);

      Animated.loop(
        Animated.timing(animatedValue, {
          duration: 2000,
          toValue: 1,
          useNativeDriver: true,
          easing: Easing.inOut(Easing.cubic),
        }),
      ).start();
    }
  }, [animatedValue, value]);

  const transform = [];

  if (value !== undefined) {
    transform.push({
      translateX: animatedValue.interpolate({
        inputRange: [0, 100],
        outputRange: [-width / 2, 0],
        extrapolate: 'clamp',
      }),
    });

    transform.push({
      scaleX: animatedValue.interpolate({
        inputRange: [0, 100],
        outputRange: [0, 1],
        extrapolate: 'clamp',
      }),
    });
  } else {
    transform.push({
      translateX: animatedValue.interpolate({
        inputRange: [0, 1],
        outputRange: [-width, width],
        extrapolate: 'clamp',
      }),
    });
  }

  const rootStyles = [
    {
      height: size,
      borderRadius: getRadius(cornerRadius, size),
    },
    styles.root,
    style,
  ];

  return (
    <View testID="root" {...props} style={rootStyles} onLayout={onLayout}>
      <Animated.View style={[styles.indicator, indicatorStyle, { transform }]} />
    </View>
  );
};

const createStyleSheet = ({ Colors }: Theme) =>
  StyleSheet.create({
    root: {
      overflow: 'hidden',
      flexDirection: 'row',
      justifyContent: 'flex-start',
      backgroundColor: Colors.gray5,
    },
    indicator: {
      width: '100%',
      height: '100%',
      backgroundColor: Colors.blue,
    },
  });

export type Props = ProgressBarProps & WithThemeProps;

export default withTheme(ProgressBar, createStyleSheet, 'ProgressBar');
