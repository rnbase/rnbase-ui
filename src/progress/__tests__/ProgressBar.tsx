import React from 'react';
import { Animated as AnimatedObj } from 'react-native';

// Note: test renderer must be required after react-native.
import TestRenderer, { act } from 'react-test-renderer';

import { Props } from '../ProgressBar';

jest.doMock('react-native/Libraries/Animated/src/Animated', () => {
  const Animated = jest.requireActual('react-native/Libraries/Animated/src/Animated');

  return {
    ...Animated,
    loop: jest.fn((value: any) => value),
    timing: jest.fn((value: any, { toValue }: any) => {
      value.setValue(toValue);

      return {
        start: jest.fn(),
      };
    }),
  };
});

/**
 * Under test
 */
const { default: ProgressBar } = jest.requireActual('../ProgressBar');

jest.mock('react-native/Libraries/Animated/src/Easing', () => ({
  out: (v: any) => `Easing.out(${v})`,
  inOut: (v: any) => `Easing.inOut(${v})`,
  cubic: 'Easing.cubic',
}));

const Animated: any = AnimatedObj;

const createElement = (props: Partial<Props>) => <ProgressBar {...props} />;

const createRenderer = (props: Partial<Props>) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  const tree = createRenderer({ value: 35 });

  act(() => {
    const root = tree.root.findByProps({ testID: 'root' });
    const event = { nativeEvent: { layout: { width: 100 } } };

    root.props.onLayout(event);
  });

  expect(tree).toMatchSnapshot();
});

it('should render as infinite progress', () => {
  const tree = createRenderer({ value: undefined });

  act(() => {
    const root = tree.root.findByProps({ testID: 'root' });
    const event = { nativeEvent: { layout: { width: 100 } } };

    root.props.onLayout(event);
  });

  expect(tree).toMatchSnapshot();
});

it('should render with custom props', () => {
  const tree = createRenderer({
    value: 35,
    size: 10,
    cornerRadius: 0,
    style: { backgroundColor: 'green' },
    indicatorStyle: { backgroundColor: 'red' },
  });

  act(() => {});

  expect(tree).toMatchSnapshot();
});

describe('animation', () => {
  const animated = { start: jest.fn() };
  const animatedLoop = { start: jest.fn() };

  Animated.timing.mockImplementation(() => animated);
  Animated.loop.mockImplementation(() => animatedLoop);

  beforeEach(() => {
    animated.start.mockClear();
    animatedLoop.start.mockClear();
  });

  it('should animate value change', () => {
    const oldValue = 0;
    const newValue = 50;
    const tree = createRenderer({ value: oldValue });

    act(() => {});

    Animated.timing.mockClear();

    act(() => {
      tree.update(createElement({ value: newValue }));
    });

    expect(Animated.timing).toBeCalledWith(expect.objectContaining({ _value: oldValue }), {
      toValue: newValue,
      duration: 200,
      useNativeDriver: true,
      easing: 'Easing.out(Easing.cubic)',
    });
    expect(animated.start).toBeCalled();
  });

  it('should animate to indeterminate', () => {
    const tree = createRenderer({ value: 50 });

    act(() => {});

    Animated.timing.mockClear();

    act(() => {
      tree.update(createElement({ value: undefined }));
    });

    expect(Animated.timing).toBeCalledWith(expect.objectContaining({ _value: 0 }), {
      toValue: 1,
      duration: 2000,
      useNativeDriver: true,
      easing: 'Easing.inOut(Easing.cubic)',
    });
    expect(Animated.loop).toBeCalledWith(animated);
    expect(animatedLoop.start).toBeCalled();
  });
});
