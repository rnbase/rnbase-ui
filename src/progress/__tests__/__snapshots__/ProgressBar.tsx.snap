// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as infinite progress 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "borderRadius": 3,
        "height": 6,
      },
      Object {
        "backgroundColor": "#e5e5ea",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "overflow": "hidden",
      },
      undefined,
    ]
  }
  testID="root"
>
  <View
    style={
      Object {
        "backgroundColor": "#007aff",
        "height": "100%",
        "transform": Array [
          Object {
            "translateX": -100,
          },
        ],
        "width": "100%",
      }
    }
  />
</View>
`;

exports[`should render normally 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "borderRadius": 3,
        "height": 6,
      },
      Object {
        "backgroundColor": "#e5e5ea",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "overflow": "hidden",
      },
      undefined,
    ]
  }
  testID="root"
>
  <View
    style={
      Object {
        "backgroundColor": "#007aff",
        "height": "100%",
        "transform": Array [
          Object {
            "translateX": -32.5,
          },
          Object {
            "scaleX": 0.35,
          },
        ],
        "width": "100%",
      }
    }
  />
</View>
`;

exports[`should render with custom props 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "borderRadius": 0,
        "height": 10,
      },
      Object {
        "backgroundColor": "#e5e5ea",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "green",
      },
    ]
  }
  testID="root"
>
  <View
    style={
      Object {
        "backgroundColor": "red",
        "height": "100%",
        "transform": Array [
          Object {
            "translateX": -0,
          },
          Object {
            "scaleX": 0.35,
          },
        ],
        "width": "100%",
      }
    }
  />
</View>
`;
